src/coreclr/nativeaot/System.Private.CoreLib/src/System/Threading/Lock.cs

```diff
@@ -244,13 +295,26 @@ private bool TryAcquireContended(int currentThreadId, int millisecondsTimeout, b
                         }
                     }
 
-                    // spinning was unsuccessful. reduce spin count.
-                    if (iteration == localSpinLimit && localSpinLimit > MinSpinLimit)
-                        _spinLimit = localSpinLimit - 1;
-
                     if (iteration++ < localSpinLimit)
                     {
-                        Thread.SpinWaitInternal(1);
+                        int newOwner = _owningThreadId;
+                        if (newOwner != 0 && newOwner != oldOwner)
+                        {
+                            ownerChanged++;
+                            oldOwner = newOwner;
+                        }
+
+                        if (canAcquire)
+                        {
+                            collisions++;
+                        }
+
+                        // We failed to acquire the lock and want to retry after a pause.
+                        // Ideally we will retry right when the lock becomes free, but we cannot know when that will happen.
+                        // We will use a pause that doubles up on every iteration. It will not be more than 2x worse
+                        // than the ideal guess, while minimizing the number of retries.
+                        // We will allow pauses up to 64~128 spinwaits, or more if there are collisions.
+                        ExponentialBackoff(Math.Min(iteration, 6) + collisions);
```

Comment:
There are two factors that inform the backoff here - iteration and collisions. 
The retry period would be normally driven by iterations. As we spin more, we reduce the rate of reads, as the comment explains. The goal is to acquire the lock within 2X of optimal wait if noone wants it, while avoiding possible interference with other threads acquiring/releasing the lock. 

When we see collisions (i.e. our interlocked operations start failing), it is a signal that we may still have more competition at the lock than we would like, so we backoff more. 
Collisions should be relatively rare - basically a "panic mode" to resolve unexpected bursts of activity.

Reply:
Should there be a cap on the collisions too similarly to the iteration, so as to avoid a perhaps rare but potentially very large backoffs?

Reply:
ExponentialBackoff has a total limit controlled by `MaxExponentialBackoffBits`.
It would be extremely unlikely to have so many collisions, but we have a limit just in case.

Reply:
Unless I'm missing something, 10 bits seems like a lot to me as a cap for one spin iteration. If we assume `SpinWait(1)` takes in the order of 10 ns, `SpinWait((1 << 10) / 2)` would take in the order of 5 us, which is quite a lot for one spin-wait iteration, despite it being rare.

Reply:
Because of double-up increases, by the time the pause reaches 5us, we have already spent 5us spinning. Basically the horribly long pauses will take horribly long time to achieve. The math kind of tells us that the limit is unnecessary, yet for many people, me included, an unbounded spin would trigger an alarm, so we need some limit. Because, who knows, we may not be considering everything ...

There is not much science behind this number though. I wanted to pick something that does not stand in the way in a typical scenario, yet protects from degenerate cases (i.e. not spinning for seconds).  Thus I picked `2^10`. I did not see it achieved in experiments (not even close), yet it is feels safer than `2^16` - another number I'd pick.

If there is any way to better estimate a better number, we could change it.

It would be interesting to get access to a 1024 core machine, or how high the core count goes these days, and see how high the collisions numbers could possibly get if trying intentionally for it (also to see if randomization helps). 
I think experimenting on 32 cores will mostly tell us that this number does not matter.